File: fudi-permission/src/test/java/com/romellfudi/permission/FudiPermisionTest.java
package com.romellfudi.permission;

import android.app.Activity;
import android.content.SharedPreferences;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.os.Build;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.invocation.InvocationOnMock;
import org.mockito.stubbing.Answer;
import org.powermock.modules.junit4.PowerMockRunner;

import java.util.ArrayList;

import static android.Manifest.permission.ACCESS_FINE_LOCATION;
import static android.Manifest.permission.CAMERA;
import static android.Manifest.permission.WRITE_EXTERNAL_STORAGE;
import static org.hamcrest.CoreMatchers.equalTo;
import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.mockito.Matchers.any;
import static org.mockito.Matchers.eq;
import static org.mockito.Mockito.verify;
import static org.powermock.api.mockito.PowerMockito.doAnswer;
import static org.powermock.api.mockito.PowerMockito.when;

/**
 * @version 1.0
 * @autor Romell Domínguez
 * @date 10/4/18
 */
@RunWith(PowerMockRunner.class)
public class FudiPermisionTest {

    @Mock
    Activity activity;

    @InjectMocks
    PermissionService permissionService = new PermissionService(activity);

    @Mock
    PermissionService.Callback callback;

    @Mock
    SharedPreferences sharedPreferences;

    @Mock
    SharedPreferences.Editor editor;

    @Mock
    PermisionServiceInterface permisionServiceInterface;

    @Captor
    ArgumentCaptor<ArrayList<String>> refusePermissionsCaptor;

    @Test
    public void requestLessM() {
        String[] permissions =
                new String[]{ACCESS_FINE_LOCATION, CAMERA, WRITE_EXTERNAL_STORAGE};
        when(permisionServiceInterface.getBuildSDK()).thenReturn(Build.VERSION_CODES.M - 1);
        permissionService.request(callback);
        verify(callback).onResponse(refusePermissionsCaptor.capture());
    }

    @Test
    public void request3Permissions3Accepted() throws PackageManager.NameNotFoundException {
        final String[] permissions =
                new String[]{ACCESS_FINE_LOCATION, CAMERA, WRITE_EXTERNAL_STORAGE};
        when(permisionServiceInterface.getBuildSDK()).thenReturn(Build.VERSION_CODES.M);
        when(activity.checkSelfPermission(any(String.class))).thenReturn(PackageManager.PERMISSION_DENIED);
        when(sharedPreferences.getBoolean(any(String.class), eq(true))).thenReturn(true);
        when(sharedPreferences.edit()).thenReturn(editor);
        when(permisionServiceInterface.getPermissions()).thenReturn(permissions);

        doAnswer(new Answer<Object>() {
            int i = 0;

            @Override
            public Object answer(InvocationOnMock invocation) throws Throwable {
                i++;
                if (i == permissions.length)
                    // 0 accepted, 1 refuse
                    callback.handler(permissions, new int[]{0, 0, 0});
                return null;
            }
        }).when(editor).apply();
        permissionService.request(callback);
        verify(callback).onResponse(refusePermissionsCaptor.capture());
    }

    @Test
    public void request3Permissions2Accepted() {
        final String[] permissions =
                new String[]{ACCESS_FINE_LOCATION, CAMERA, WRITE_EXTERNAL_STORAGE};
        when(permisionServiceInterface.getBuildSDK()).thenReturn(Build.VERSION_CODES.M);
        when(activity.checkSelfPermission(any(String.class))).thenReturn(PackageManager.PERMISSION_DENIED);
        when(sharedPreferences.getBoolean(any(String.class), eq(true))).thenReturn(true);
        when(sharedPreferences.edit()).thenReturn(editor);
        when(permisionServiceInterface.getPermissions()).thenReturn(permissions);
        doAnswer(new Answer<Object>() {
            int i = 0;

            @Override
            public Object answer(InvocationOnMock invocation) throws Throwable {
                i++;
                if (i == permissions.length)
                    // 0 accepted, 1 refuse
                    callback.handler(permissions, new int[]{0, 1, 0});
                return null;
            }
        }).when(editor).apply();

        permissionService.request(callback);
        verify(callback).onResponse(refusePermissionsCaptor.capture());

        ArrayList<String> refusePermissions = refusePermissionsCaptor.getValue();
        assertThat(refusePermissions.get(0), is(equalTo(CAMERA)));
    }

    @Test
    public void request3Permissions3Reject() {
        final String[] permissions =
                new String[]{ACCESS_FINE_LOCATION, CAMERA, WRITE_EXTERNAL_STORAGE};
        when(permisionServiceInterface.getBuildSDK()).thenReturn(Build.VERSION_CODES.M);
        when(activity.checkSelfPermission(any(String.class))).thenReturn(PackageManager.PERMISSION_DENIED);
        when(sharedPreferences.getBoolean(any(String.class), eq(true))).thenReturn(true);
        when(sharedPreferences.edit()).thenReturn(editor);
        when(permisionServiceInterface.getPermissions()).thenReturn(permissions);
        doAnswer(new Answer<Object>() {
            int i = 0;

            @Override
            public Object answer(InvocationOnMock invocation) throws Throwable {
                i++;
                if (i == permissions.length)
                    // 0 accepted, 1 refuse
                    callback.handler(permissions, new int[]{1, 1, 1});
                return null;
            }
        }).when(editor).apply();

        permissionService.request(callback);
        verify(callback).onResponse(refusePermissionsCaptor.capture());

        ArrayList<String> refusePermissions = refusePermissionsCaptor.getValue();
        assertThat(refusePermissions.get(0), is(equalTo(ACCESS_FINE_LOCATION)));
        assertThat(refusePermissions.get(1), is(equalTo(CAMERA)));
        assertThat(refusePermissions.get(2), is(equalTo(WRITE_EXTERNAL_STORAGE)));
    }
}

File: fudi-permission-kotlin/src/test/java/com/romellfudi/permission/FudiPermisionTest.kt
package com.romellfudi.permission

import android.Manifest.permission.*
import android.app.Activity
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.os.Build
import android.preference.PreferenceManager
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.whenever
import io.mockk.*
import io.mockk.impl.annotations.MockK
import org.hamcrest.CoreMatchers.`is`
import org.hamcrest.CoreMatchers.equalTo
import org.hamcrest.MatcherAssert.assertThat
import org.junit.Before
import org.junit.Test
import org.mockito.Mock
import org.powermock.core.classloader.annotations.PrepareForTest
import java.util.*

/**
 * @version 1.0
 * @autor Romell Domínguez
 * @date 10/4/18
 */
@PrepareForTest(PreferenceManager::class)
class FudiPermisionTest {

    @MockK
    lateinit var activity: Activity

    @MockK
    lateinit var permissionService: PermissionService

    @MockK
    lateinit var callback: PermissionService.Callback

    @MockK
    lateinit var sharedPreferences: SharedPreferences

    @Mock
    internal var editor: SharedPreferences.Editor = mock()

    @MockK
    lateinit var permisionServiceInterface: PermisionServiceInterface

    val refusePemissions = slot<ArrayList<String>>()

    val permissions = arrayOf(ACCESS_FINE_LOCATION, CAMERA, WRITE_EXTERNAL_STORAGE)

    @Before
    fun setUp() {
        MockKAnnotations.init(this, relaxUnitFun = true)
        mockkStatic(PreferenceManager::class)
        every { PreferenceManager.getDefaultSharedPreferences(any()) } returns sharedPreferences
        permissionService = PermissionService(activity)
        permissionService.permisionServiceInterface = permisionServiceInterface
        i = 0
        every { permisionServiceInterface.buildSDK } returns Build.VERSION_CODES.M
        every { activity.checkSelfPermission(any()) } returns PackageManager.PERMISSION_DENIED
        every { sharedPreferences.getBoolean(any(), any()) } returns true
        every { sharedPreferences.edit() } returns editor
    }

    @Test
    fun requestLessM() {
        every { permisionServiceInterface.buildSDK } returns Build.VERSION_CODES.M - 1
        permissionService.request(permissions, callback)
        verify { callback.onFinally() }
    }

    companion object {
        var i = 0
    }

    @Test
    fun request3Permissions3Accepted() {
        whenever(editor.apply()).thenAnswer {
            i++
            if (i == permissions.size)
            // 0 accepted, 1 refuse
                PermissionService.handler(callback, intArrayOf(0, 0, 0), permissions)
        }
        permissionService.request(permissions, callback)

        verify { callback.onFinally() }
    }

    @Test
    fun request3Permissions2Accepted() {
        whenever(editor.apply()).thenAnswer {
            i++
            if (i == permissions.size)
            // 0 accepted, 1 refuse
                PermissionService.handler(callback, intArrayOf(0, 1, 0), permissions)
        }
        permissionService.request(permissions, callback)
        verify { callback.onRefuse(capture(refusePemissions)) }

        val refusePermissions = refusePemissions.captured
        assertThat(refusePermissions[0], `is`(equalTo(CAMERA)))
    }


    @Test
    fun request3Permissions3Reject() {
        whenever(editor.apply()).thenAnswer {
            i++
            if (i == permissions.size)
            // 0 accepted, 1 refuse
                PermissionService.handler(callback, intArrayOf(1, 1, 1), permissions)
        }
        permissionService.request(permissions, callback)
        verify { callback.onRefuse(capture(refusePemissions)) }

        val refusePermissions = refusePemissions.captured
        assertThat(refusePermissions[0], `is`(equalTo(ACCESS_FINE_LOCATION)))
        assertThat(refusePermissions[1], `is`(equalTo(CAMERA)))
        assertThat(refusePermissions[2], `is`(equalTo(WRITE_EXTERNAL_STORAGE)))
    }
}